README.md
# gethamquestions
Parse the official Amateur Radio FCC question pools and get the Element Questios in a JSON object

## Installation
Download the getquestions.py file. 

## Overview
With the advent of accessible AI infrastructure, Amateur Radio seems a perfect domain in which to experiment with value add functions that AI API's might provide.
To get started though, requires accurate question pools.  Fortunately these are publically available from [http://www.ncvec.org/page.php?id=338](http://www.ncvec.org/page.php?id=338).
You can run gethamquestions.py specifying the docx file as input.  It will create a JSON file as output, and also a text version (UTF-8 encoding) of the docx file.  If you prefer, you can specifiy a text file as input (recommended UTF-8 encoding.)

A javascript class is also provided, elementpool.js (class ElementPool) that supports the method "getQuestionsByIds".  Additional methods will be added as use cases emerge.

The next phase of development will experiment with AI input on topics, and explainations of the questions and answers. Also provide full package support.

A rudimentary deployment of gethamquestions can be seen at [projectnotions.com - Technician Ham Class](https://projectnotions.com/blog1/2023/05/technician-ham-class)

## Usage

```python
python gethamquestions.py "C:\Users\kb\onedrive\HamTest\QuestionPools\Technician Pool and Syllabus 2022-2026 Public Release Errata March 7 2022.utf8.txt"

# returns json file, txt file (if the input was docx), and summary (abbreviated below)
Info I9999 C:\Users\kenwb\onedrive\HamTest\QuestionPools\Technician Pool and Syllabus 2022-2026 Public Release Errata March 7 2022.utf8.txt 0 nond
timestamp = 2023-06-22 16:13:45.848043
JSON written to Element2.json
*** Summary - Element 2 - self.cur_element.timestamp (Technician Class) 2022-2026 ***
Subelements: 10
  Sub element: T1, Groups: 6
    Group: A, questions: 11
    Group: B, questions: 12
    Group: C, questions: 11
    Group: D, questions: 11
    Group: E, questions: 11
    Group: F, questions: 11
  Sub element: T2, Groups: 3

 etc. etc.

   Sub element: T0, Groups: 3
    Group: A, questions: 12
    Group: B, questions: 11
    Group: C, questions: 13
Subelements: 10, groups: 35, questions: 411 
```
The javascript class ElementPool is provided for methods of getting guestions based on various criteria.

## Output

A JSON file is created with the name of "ElementX.json" where X is 2, 3, or 4.  A sample is below:

```javascript
{
  "filename": "C:\\Users\\kenwb\\onedrive\\HamTest\\QuestionPools\\Technician Pool and Syllabus 2022-2026 Public Release Errata March 7 2022.utf8.txt",
  "timestamp": "2023-06-22 16:13:45.848043",
  "elem": "2",
  "elname": "Technician",
  "yrvalid": {
    "begin": "2022",
    "end": "2026"
  },
  "effective": {
    "begin": "7/01/2022",
    "end": "6/30/2026"
  },
  "subelements": [
    {
      "elem": "2",
      "sub_el": "T1",
      "description": "COMMISSION'S RULES",
      "numq": "6",
      "numg": "6",
      "groups": [
        {
          "subelement": "T1",
          "group_id": "A",
          "description": "Purpose and permissible use of the Amateur Radio Service; Operator/primary station license grant; Meanings of basic terms used in FCC rules; Interference; RACES rules; Phonetics; Frequency Coordinator",
          "topics": [
            "Purpose and permissible use of the Amateur Radio Service",
            "Operator/primary station license grant",
            "Meanings of basic terms used in FCC rules",
            "Interference",
            "RACES rules",
            "Phonetics",
            "Frequency Coordinator"
          ],
          "questions": [
            {
              "subelement": "T1",
              "group": "A",
              "num": "01",
              "qid": "T1A01",
              "text": "Which of the following is part of the Basis and Purpose of the Amateur Radio Service?",
              "correct": "C",
              "figure": "",
              "fcc": "[97.1]",
              "answers": [
                "A. Providing personal radio communications for as many citizens as possible",
                "B. Providing communications for international non-profit organizations",
                "C. Advancing skills in the technical and communication phases of the radio art",
                "D. All these choices are correct"
              ],
              "topics": [
                "Purpose and permissible use of the Amateur Radio Service"
              ]
            },
```


CHANGES.txt
-------------------------------------------------------------------------------------
gethamquestions Package Changes Log   
-------------------------------------------------------------------------------------

Release Summary:

0.next 
     2023-09-02 added experiment directory for notes, code, cheatsheets

0.1, June 27, 2023: 
     Initial pre-release.

Details of each release follow, from most recent to oldest.
-------------------------------------------------------------------------------------

Release 0.1, June 27, 2023

Working release with basic function (read Element pool text file).
Preliminary packing on github.

-------------------------------------------------------------------------------------

experiments/onelinecodes.py
# from https://medium.com/@hafizabdullaha93/20-extremely-useful-single-line-python-codes-de44928a1a3b
# One line codes:
# 1. One-line For Loop
# 2. One-Line while Loop
# 3. One-Line If Else Statement
# 4. Merging dictionaries in one line
# 5. One Line Function
# 6. Recursion in one line
# 7. Filtering an array in one line
# 8. Exception Handling in One Line
# 9. Converting a List to a Dictionary in One Line
# 10. One-Line Multiple Variables
# 11. Swapping values in one line
# 12. One-Line Sorting
# 13. Reading a file in one line
# 14. One-Line Class
# 15. One-Line semicolon
# 16. One Line Print
# 17. One Line Map Function
# 18. Deleting the Mul elements from the first row of the list
# 19. One Line Print Pattern
# 20. Find Primes in a range in a single line of code

#For loop in one line
mylist = [200, 300, 400, 500]
#Single line For loop
result = [] 
for x in mylist: 
    if x > 250: 
        result.append(x) 
print(result) # [300, 400, 500]
#One-line code way
result = [x for x in mylist if x > 250] 
print(result) # [300, 400, 500]

# One line While Loop
#Method 1 Single Statement   
while True: print(1) #infinite 1  
#Method 2 Multiple Statements  
x = 0   
while x < 5: print(x); x= x + 1 # 0 1 2 3 4 5

#if Else In a single line.  
#Example 1 if else  
print("Yes") if 8 > 9 else print("No") # No  
#Example 2 if elif else   
E = 2   
print("High") if E == 5 else print("数据STUDIO") if E == 2 else   
print("Low") # Data STUDIO   
   
#Example 3 only if  
if 3 > 2: print("Exactly") # Exactly

# Merging dictionaries in one line  
d1 = { 'A': 1, 'B': 2 }   
d2 = { 'C': 3, 'D': 4 }  
#Method1   
d1.update(d2)   
print(d1) # {'A': 1, 'B': 2, 'C': 3, 'D': 4}  
#Method2   
d3 = {**d1, **d2}   
print(d3) # {'A': 1, 'B': 2, 'C': 3, 'D': 4}

#Function in one line  
#Method1
def fun(x): return True if x % 2 == 0 else False   
print(fun(2)) # False  
#Method2
fun = lambda x : x % 2 == 0   
print(fun(2)) # True   
print(fun(3)) # False

# Single-line recursion 
#Fibonaci Single-line recursion example  
def Fib(x): return 1 if x in {0, 1} else Fib(x-1) + Fib(x-2)  
print(Fib(5)) # 8  
print(Fib(15)) # 987

# Filtering arrays in a single line  
mylist = [2, 3, 5, 8, 9, 12, 13, 15]  
#Normal way  
result = []   
for x in mylist:   
    if x % 2 == 0:   
        result.append(x)  
print(result) # [2, 8, 12]  
# One-line method  
result = [x for x in mylist if x % 2 == 0]   
print(result) # [2, 8, 12]

# Exception handling in one line  
#Original method  
try:  
    print(x)  
except:  
    print("Error")  
#Single line way  
exec('try:print(x) \nexcept:print("Error")') # Error

# Dictionary in one line  
mydict = ["John", "Peter", "Mathew", "Tom"]  
mydict = dict(enumerate(mydict))  
print(mydict) # {0: 'John', 1: 'Peter', 2: 'Mathew', 3: 'Tom'}

#Multiple variable assignments in one line.  
#Single-line method  
x = 5   
y = 7   
z = 10   
print(x , y, z) # 5 7 10  
#Single line way  
a, b, c = 5, 7, 10   
print(a, b, c) # 5 7 10

#Swap values in one line  
#Single-line method 
v1 = 100  
v2 = 200  
temp = v1  
v1 = v2   
v2 = temp  
print(v1, v2) # 200 100  
# One-line value swapping 
v1, v2 = v2, v1   
print(v1, v2) # 200 100

# Sort in one line  
mylist = [32, 22, 11, 4, 6, 8, 12]   
# Method1
mylist.sort()   
print(mylist) # # [4, 6, 8, 11, 12, 22, 32]  
print(sorted(mylist)) # [4, 6, 8, 11, 12, 22, 32]

#Reading a file in one line  
#Single-line method 
with open("data.txt", "r") as file:   
    data = file.readline()   
    print(data) # Hello world  
#Single line way  
data = [line.strip() for line in open("data.txt","r")]   
print(data) # ['hello world', 'Hello Python']

# One-line class  
#Regular way  
class Emp:   
    def __init__(self, name, age):   
        self.name = name   
        self.age = age  
        emp1 = Emp("a44", 22)   
print(emp1.name, emp1.age) # 
#Single line way  
#Method 1 Lambda with Dynamic Attributes 
  
Emp = lambda:None; Emp.name = "a44"; Emp.age = 22  
print(Emp.name, Emp.age) # 
  
#Method 2 
from collections import namedtuple  
Emp = namedtuple('Emp', ["name", "age"]) ("a44", 22)   
print(Emp.name, Emp.age)

# One-line semicolon 
# exsample 1   
a = "Python"; b = "Programming"; c = "languages"; print(a, b, c)  
# print
# Python Programming languages

# One line print
#Single-line method
for x in range(1, 5):  
    print(x) # 1 2 3 4  
#Single line way  
print(*range(1, 5)) # 1 2 3 4  
print(*range(1, 6)) # 1 2 3 4 5

#One line map function 
print(list(map(lambda a: a + 2, [5, 6, 7, 8, 9, 10])))  
# print
# [7, 8, 9, 10, 11, 12]

# Deleting the Mul elements from the first row of the list
mylist = [100, 200, 300, 400, 500]  
del mylist[1::2]  
print(mylist) # [100, 300, 500]

# One line print pattern#   
# Single-line method  
for x in range(3):  
    print('😀')  
# print
# 😀 😀 😀  
#Single line way  
print('😀' * 3) # 😀 😀 😀   
print('😀' * 2) # 😀 😀   
print('😀' * 1) # 😀

# Find primes in a range in a single line of code
print(list(filter(lambda a: all(a % b != 0 for b in range(2, a)),  
                  range(2,20))))  
# print
# [2, 3, 5, 7, 11, 13, 17, 19]



gethamquestions/elementpool.js
// gethamquestions - Amateur Radio Exam Preparation
// @source: https://github.com/projectnotions/gethamquestions
// Copyright (C) 2023 KWBoyd
// This program is free software; you can redistribute it and/or modify it under the terms of the 
// GNU General Public License as published by the Free Software Foundation; 
// either version 2 of the License, or (at your option) any later version.  
// elementpool.js for gethamquestions (2023-07-21)
//
/** Class containing the Amateur Radio Element Question Pool
 * Generated from the question pooly by gethamquestions.py
 * More methods will be added as development proceeds
*/
//console.log('elementpool2.js begin')
//
/*********************************************/
/*** WPCode snippet name: elementpool.js   ***/
/*********************************************/
class ElementPool {
	/**
     * Create the ElementPool with an precreated object (see gethamquestions.py)
	 * or use the method getElementByFileName to load an object
     * @param {object} element - element pool object, or null
	 * @property {object} element - an element pool object, created by gethamquestions.py
	 * @property {function} getFileSuccess - a function called by getElementByFileName on success
	 * @property {function} getFileFail - a function called by GetElementByFileName on fail
     */
	constructor(element=null) {
		this.element = element;
		this.getFileSuccess =null;
		this.getFileFail = null
	}

	EpSuccess(json) {
		this.element = json;
	}

	EpFail(evt) {
		this.element = '';
	}
	/**
	 * Get an Element from a file.  The element is in JSON format.  
	 * gethamquestions.py can produce the file
	 * 
	 * @param {string} fileUrl - The url of the file.  Relative or absolute
	 * @param {function} [success] - function to be called if loading the file is successfull
	 * @param {function} [fail] - function to be called if loading the file fails
	 */
    getElementByFileName(fileUrl, success, fail) {
		debug(`getElementByFileName()`, 'Enter', `${fileUrl}`);
		if (typeof success == 'function') {
			this.getFileSuccess = success;
		}
		if (typeof fail == 'function') {
			this.getFileFail = fail;
		}
		let self = this;
		this.element = '';
		jQuery.getJSON(fileUrl, function(json) {	
		    self.EpSuccess(json);
			if (self.getFileSuccess) {
				self.getFileSuccess(json);
			}

		})
		    .fail(function(evt) {
                self.EpFail(evt);
				if (self.getFileFail) {
					self.getFileFail(evt);
				}

			})
		    //.always(function(evt) {
			//	self.EpAlways(evt);
			//})  
			debug(`getElementByFileName()`, 'Exit', `${fileUrl}`);     
    }
	/**
	 * Get Questions by their ID, i.e. T1A01, T3C04, etc.
	 * 
	 * @param {Array} qIds - An array with the question ids
	 * @param {string} [options=''] options - 'strip-answer-prefix' - Strip 'Question T1A01'
	 * @return [[{questions}], [errorMessages]] - returns question array and Msg array
	 */
	//  /***  @enum
	//  *         where Question has the properties:
	//  *         .qid - the question id, i.e. 'T1A01'
	//  *         .figure - str with the name of the associated figure/diagram,
	//  *         .correct - str, on of ['A', 'B', 'C', 'D'] that is the correct answer,
	//  *         .title - str, the question in this format: 'Question #T1A03 What is the ...',
	//  *         .answers - an array of str containing the answer strings, i.e. ,
	//  *            A. All of these answers are correct ,
	//  *            Note: if options has the string 'strip-answer-prefix' the prefix is removed. ,
    //  */
	getQuestionsByIds(qIds, options='') {
		debug(`getQuestionsById()`, 'Enter', `${qIds}`);
		let resultMsgs = [];
		if (!this.element) {
			resultMsgs.push("this.element is empty")
			return [[''], resultMsgs];
		}
		if (!Array.isArray(qIds)) {
			resultMsgs.push("input is not an array")
			return [[''], resultMsgs];
		}
		let result = [];
		let e = this.element;
		let se = e.subelements;
		for (let seI=0; seI<se.length; seI++) {
			//debug('getQuestionsByIds()', 'Info', `Subelement = "${se[seI].sub_el}", description ="${se[seI].description}"`);
			let g = se[seI].groups;
			for (let gI=0; gI<g.length; gI++) {
				//debug('getQuestionsByIds()', 'Info', `    Group = "${g[gI].group_id}", description ="${g[gI].description}"`);
				let q = g[gI].questions;
				for (let qI=0; qI<q.length; qI++) {
					//debug('getQuestionsByIds()', 'Info', `        Question = "${q[qI].qid}", text ="${q[qI].text}"`);
					let indx = qIds.indexOf(q[qI].qid);
					if (indx != -1)	{
						let question = {};
						question.qid = q[qI].qid;
						question.figure = q[qI].figure;
						if (question.figure) {
							//debug('getQuestionsByIds()', 'Info', `        Question = "${q[qI].qid}", figure ="${q[qI].figure}"`);	
						}
						let correct = ['A', 'B', 'C', 'D'].indexOf(q[qI].correct);
						question.title = `Question #${q[qI].qid} ${q[qI].text}`;
						let start = 0;
						if (options.indexOf('strip-answer-prefix') != -1) {
						   start = 3;	
						}
						question.answers = [];
						for (let i=0; i<q[qI].answers.length; i++) {
							question.answers[i] = q[qI].answers[i].substr(start);
						}	
						question.correct = correct;
						result[indx] = question;
						resultMsgs[indx] = 'question found';
					}
				}
			}
		}
		for (let i=0; i<result.length; i++) {
		    if (!result[i]) {
				result[i] = {};
				let errMsg = `result is empty at index ${i}`
				resultMsgs[i] = errMsg;
				debug('getQuestionById()', 'Error', errMsg);
			}	
		}
		debug(`getQuestionsById()`, 'Exit', `${qIds}`);
		return [result, resultMsgs];
	}
}


gethamquestions/gethamquestions.py
#-*- coding: utf-8 -*-
"""
Process the Amateur Radio Element Question Pool and return a JSON representation
The following question pools have been tested:

Classes:

    Element
    Subelement
    Group
    Question

Functions:
    _parse_line
    get_element

Misc variables:
    REGEX_DICT

References
    https://www.geeksforgeeks.org/read-a-file-line-by-line-in-python/
    https://www.vipinajayakumar.com/parsing-text-with-python/
    stackoverflow.com/questions/55933956/what-does-a-star-asterisk-do-in-f-string/55934921
    https://docs.python.org/3/howto/unicode.html#introduction-to-unicode
    https://etienned.github.io/posts/extract-text-from-word-docx-simply/

Notes
    pylint is used for style checking
    see function _parse_line() notes for Question Pool anomolies

Change Log
    2023-07-05 v12 - processed group description string into topics and subtopics
    2023-07-04 v11 - problem with very long topic in T1C01
    2023-06-29 v10 - adding openai routine to identify topics for question; will be separate process
    2023-06-30 v10 - tried to resolve use of u\uf0b4 for "x", could not fix, 7/5 fixed
    2023-06-29 v09 - fixed unusal characters for +', etc
    2023-06-28 v08 - added case for "E3B08 (DELETED)"
    2023-06-27 v07 - removed match statement for wls python 3.8
    2023-06-22 v06 - cleaned up pylint
    2023-06-13 v05 - Added source file name to
    2023-05-24 v04 - Work for all elements as of now.  Added Question id attribute
    2023-05-22 v03 - Wrks for element 2 and element 3. State base for better error detection
    2023-05-21 v02 = Works for Question Pool only. Not for total file. Is not state based.
"""

import re
import json
import sys
import datetime
import os
import os.path
#import docx
import magic
#import zipfile
from gethamexternalfunctions import get_docx_text
from gethamelementclasses import Element, Subelement, Group, Question
from gethamquestionclasses import msg, State, Filelines

# set up regular expressions
# use https://regexper.com to visualise these if required
REGEX_DICT = {
    """
    Examples:
        subelement
            element 4 SUBELEMENT E1 - COMMISSION RULES [6 Exam Questions - 6 Groups]
    """

    'element_onel' : re.compile(
        r'(?P<begin>\d\d\d\d)-(?P<end>\d\d\d\d)\s+'
        r'(?P<elname>[a-zA-z]+)\s+Class.*FCC Element '
        r'(?P<elnum>\d)\sQuestion Pool\s-\sEffective\s(?P<effbegin>\w+\s\d+,\s\d\d\d\d).*'),
    'el_name'    : re.compile(
        r'(?P<yrbegin>\d\d\d\d)-(?P<yrend>\d\d\d\d)\s+(?P<elname>[a-zA-z]+)\s+Class.*'),
    'el_effective': re.compile(r'Effective (?P<begin>\d+\/\d+/\d+)...(?P<end>\d+\/\d+\/\d+)'),
    'el_eff_short': re.compile(r'Effective (?P<begin>[JFMASOND].*\s\d+\,\s\d\d\d\d).*$'),
    'el_num'   : re.compile(r'FCC Element (?P<elnum>\d)\sQuestion Pool\s*$'),
    'subelement': re.compile(r'SUBELEMENT (?P<subelement>[TGE]\d)\s.?\s?(?P<description>.*)'\
                  r'-?\s?\[(?P<numq>\d+)\s[Ee]xam\s[Qq]uestions?\s.*\s(?P<numg>\d+)\s[Gg]roups?\]'),
    'subelement2': re.compile(r'(?P<subelement>[TGE]\d)\s.\s(?P<description>.*)'\
                   r'\s.?\[(?P<numq>\d+)\sExam\sQuestions\s.\s(?P<numg>\d+)\s[Gg]roups\].*'),
    'group'     : re.compile(
        r'(?P<subelem>[TGE]\d)(?P<group_id>[A-H])\s-?\s?(?P<description>.*)'),
    'question'  : re.compile(
        r'(?P<subelem>[TGE]\d)(?P<group>[A-H])(?P<qnum>\d\d)\s?\((?P<ans>[A-D])\)\s?(?P<fcc>.*)'),
    #'question'  : re.compile(
    # r'(?P<subelem>T\d)(?P<group>[A-F])(?P<qnum>\d\d).+\((?P<ans>[A-D])\)\s?(?P<fcc>.*)'),
    'removed'   :
        re.compile(r'(?P<subelem>[TGE]\d)(?P<group>[A-F])(?P<qnum>\d\d)\s? Question Removed.?'),
    'removed2'  : re.compile(
        r'(?P<subelem>[TGE]\d)(?P<group>[A-F])(?P<qnum>\d\d)\s? \(DELETED\).?'),
    'end'      : re.compile(r'~~~~?.*~~~~?'),
    'blank'    : re.compile(r'~~'),
}

def _parse_line(line, pool_state):
    """
    Do a regex search against all defined regexes and
    return the key and match result of the first matching regex

    Question heading formats:
      Element 2: 2022-2026: 2022-2026 Technician Class
                          : FCC Element 2 Question Pool
                          : Effective 7/01/2022 - 6/30/2026
      Element 3: 2019-2023: General Class - FCC Element 3 Question Pool - Effective July 1, 2019
      Element 3: 2023-2027: 2023-2027 General Class
                          : FCC Element 3 Question Pool
                          : Effective 7/01/2023 - 6/30/2027
      Element 4: 2020-2024: 2020-2024 Extra Class
                          : FCC Element 4 Question Pool
                          : Effective July 1, 2020
    Anomolies:
      Element 4: E3B08 (DELETED)
      Element 4: 2020-2024: E7C09 (D) - Missing a ~~ line at the end of the question
                 (~~ at end of answer D, line wrap disguises error in word)
      Element 4: 2020-2024: E8C10 (C) - Missing a ~~ line at the end of the question
      Element 4: file: 2020ExtraClassPoolJan22.txt
                  ----------------------------------------------------------------------
                  E2B08 (A)
                    What technique allows commercial analog TV receivers to be used for fast-scan
                        TV operations on the 70 cm band?
                    A. Transmitting on channels shared with cable TV
                    B. Using converted satellite TV dishes
                    C. Transmitting on the abandoned TV channel 2
                    D. Using USB and demodulating the signal with a computer sound card

                    ~~
                  ----------------------------------------------------------------------
                  ~~~~End of question pool text~~~~
                  ----------------------------------------------------------------------
    """
    if not line:
        return 'end', ''
    for key, regex in REGEX_DICT.items():
        match = regex.search(line)
        #print(key + ", match='" + match + "'")
        if match:
            if key == 'element_onel':
                # print('element_one1 match')
                pool_state.el_name = match.group('elname')
                pool_state.el_effective = {'begin' : match.group('effbegin'), 'end' : ''}
                pool_state.el_yrvalid = {'begin' : match.group('begin'), \
                                       'end'   : match.group('end')}
                pool_state.el_num = match.group('elnum')
                key = 'element'
            if key == 'el_name':
                pool_state.el_name = match.group('elname')
                pool_state.el_yrvalid = {'begin' : match.group('yrbegin'), \
                                       'end'   : match.group('yrend')}
                key = 'data'
            if key == 'el_effective':
                pool_state.el_effective = {'begin' : match.group('begin'), \
                                         'end'   : match.group('end')}
                key = 'data'
                if pool_state.el_num:
                    key = 'element'
            if key == 'el_eff_short':
                pool_state.el_effective = {'begin' : match.group('begin'), \
                                         'end'   : ''}
                key = 'data'
                if pool_state.el_num:
                    key = 'element'

            if key == 'el_num':
                pool_state.el_num = match.group('elnum')
                key = 'data'
            if key == 'subelement2':
                key = 'subelement'

            return key, match
    # if there are no matches
    return None, None

def get_file_type(file_name):
    """
    Determine the file type and return:
    - "ASCII text" if it is a text file
    - "Microsoft Word" if it is a docx
    - "anything else" - not ASCII text or Microsoft Word

    """

    tokens = magic.from_file(file_name).split()
    result = ''
    sep = ''
    i = 0
    while (i < len(tokens)) and (i < 2):
        result += sep + tokens[i]
        sep = ' '
        i += 1
    return result

def get_file(file_name):
    """
    Read a file and return an iterable object list of all lines

    """

    file_type = get_file_type(file_name)
    if file_type in ('ASCII text', 'UTF-8 Unicode'):
        try:
            with open(file_name, 'r', encoding='UTF-8') as file:
                lines = file.readlines()
            return lines

        except IOError as err:
            print(f'I/O error({err.errno}): {err.strerror} for: "{file_name}"')
            return ''

        except: #handle other exceptions such as attribute errors
            print("Unexpected error:", sys.exc_info()[0])
            return ''
    elif file_type == 'Microsoft Word':
        lines = get_docx_text(file_name)
        msg('Info', 'I400', 'get_file(' + file_name + ')' +
            'returned len(lines)= ' + str(len(lines)))
        return lines
    else:
        msg('Error', 'E401', 'Unknown file type "' + file_type + '"')
        return ''

def fix_line(line):
    """
    Replace common unicode characters with the ASCII version

    """
    line = line.replace(u"\u2013", '-')
    line = line.replace(u"\u2019", "'")
    line = line.replace(u"\u2018", "'")
    line = line.replace(u"\u201c", '"')
    line = line.replace(u"\u201d", '"')
    # \uF0B4 is a symbol in a symbol table
    # these are removed in get_docx_text
    # line = line.replace(u"\uf0b4", 'x')
    return line

def read_fline(filelines, skip_blank=True):
    """
    Read the next non-blank line of the iterable

    """

    try:
        line, index = next(filelines)
        if not line.isascii():
            msg('Warning', 'W403', 'line has non-ascii characters', index, json.dumps(line))
            line = fix_line(line)
        while line and (len(line.strip()) == 0 and skip_blank):
            line, index = next(filelines)
            if not line.isascii():
                msg('Warning', 'W403', 'line has non-ascii characters', index, json.dumps(line))
                line = fix_line(line)
        return line, index

    except StopIteration:
        line = ''
        return line, len(filelines)

def get_element_pool(file_name):
    """
    Extract the element pool from the source file

    """
    #State __init__(self, state, cur_element, cur_subelement, cur_group):
    #state.elname = ''
    file_lines = get_file(file_name)
    pool_state = State('initial', None, None, None, file_lines)
    file_lines = Filelines(file_lines)  # convert to Filelines iterable
    while pool_state.state != 'end':
        line, count = read_fline(file_lines)
        key, match = _parse_line(line, pool_state)
        begin_state = pool_state.state

        if key in ('removed', 'removed2', 'blank'):
            continue
        #match pool_state.state:
            #case 'initial':
        if pool_state.state == 'initial':
                #match key:
                    #case 'element':
            if key == 'element':
                # Can be ended by end of input
                # Not currently allowed more than one element, so it can't end anything else
                subelements = []
                timestamp = datetime.datetime.now()
                pool_state.cur_element = Element(pool_state.el_num, pool_state.el_name, \
                    pool_state.el_yrvalid, pool_state.el_effective, subelements, \
                    timestamp, file_name, get_file_type(file_name))
                pool_state.state = 'element'
                    #case 'end':
            elif key == 'end':
                msg('Error', 'E001', 'Premature end', count, line)
                pool_state.state = 'end'
                    #case 'data':
            elif key == 'data':
                msg('Debug', 'D002', 'Data line', count, line)
                    #case _:
            else:
                msg('Debug', 'D004', 'Line b4 Element', count, line)

            #case 'element':
        elif pool_state.state == 'element':
                #match key:
                    #case 'subelement':
            if key == 'subelement':
                # Can be ended by subelement, end of input
                # Can end group, subelement
                #                        pool_state.close_group()
                #                        pool_state.close_subelement()
                # New subelement
                description = match.group('description').strip().rstrip('-').strip()
                pool_state.cur_subelement = \
                    Subelement(pool_state.cur_element.elem, match.group('subelement'), \
                                description, match.group('numq'), match.group('numg'), [])
                pool_state.state = 'subelement'
                    #case 'end':
            elif key == 'end':
                msg('Error', 'E003', 'Unexpected "end"', count, line)
                pool_state.state = 'end'
                    #case _:
            else:
                msg('Error', 'E004', '{key} from {pool_state.state}', count, line)
                pool_state.state = 'end'
            #case 'subelement':
        elif pool_state.state == 'subelement':
                #match key:
                    #case 'group':
            if key == 'group':
                # Can be ended by subelement, group, end of input
                # Can end group
                pool_state.close_group()
                # New group
                subelem = match.group('subelem')
                group_id = match.group('group_id')
                description = match.group('description')
                description = description.strip().rstrip('-').strip()
                questions = []
                pool_state.cur_group = Group(subelem, group_id, description, questions)
                pool_state.state = 'group'
                    #case 'end':
            elif key == 'end':
                msg('Error', 'E004', 'Unexpected "end"', count, line)
                pool_state.state = 'end'
                    #case _:
            else:
                msg('Error', 'E007', '{key} from {pool_state.state}', count, line)
                msg('Error', 'D008', f'{begin_state}:{pool_state.state}', count, line)
                pool_state.state = 'end'
            #case 'group':
        elif pool_state.state == 'group':
                #match key:
                    #case 'group':
            if key == 'group':
                # Can be ended by subelement, group, end of input
                # Can end group
                pool_state.close_group()
                # New group
                subelem = match.group('subelem')
                group_id = match.group('group_id')
                description = match.group('description')
                questions = []
                pool_state.cur_group = Group(subelem, group_id, description, questions)
                pool_state.state = 'group'
                    #case 'subelement':
            elif key == 'subelement':
                # Can be ended by subelement?, end of input
                # Can end group, subelement?
                pool_state.close_group()
                pool_state.close_subelement()
                sub_el = match.group('subelement')
                description = match.group('description')
                # couldn't get regex to elminiate final - in some cases
                description = description.strip().rstrip('-').strip()
                numq = match.group('numq')
                numg = match.group('numg')
                groups = []
                pool_state.cur_subelement = \
                    Subelement(pool_state.cur_element.elem, sub_el, \
                                description, numq, numg, groups)
                pool_state.state = 'subelement'
                    #case 'question':
            elif key == 'question':
                msg('Debug', 'D002', f'{begin_state}:{pool_state.state}', count, line)
                subelem = match.group('subelem')
                group = match.group('group')
                qnum = match.group('qnum')
                qid = f'{subelem}{group}{qnum}'
                ans = match.group('ans')
                fcc = match.group('fcc')
                # Get question lines from the file
                text, count = read_fline(file_lines)      # read line 1 Question
                figure = ''
                regex_figure = re.compile(r'(^|\s)[fF]igure\s+(?P<fig>[TGE]\d?-\d+).?')
                match = regex_figure.search(text)
                if match:
                    figure = match.group('fig')
                answers = []

                line, count = read_fline(file_lines)      # read line 2 Ans A.
                answers.append(line.strip())
                line, count = read_fline(file_lines)      # read line 3 Ans B.
                answers.append(line.strip())
                line, count = read_fline(file_lines)      # read line 4 Ans C.
                answers.append(line.strip())
                line, count = read_fline(file_lines)      # read line 5 Ans D.
                answers.append(line.strip())
                # ignore this line, ~~ at end of question
                # Read line 6 Question End ~~, don't skip blank lines
                line, count = read_fline(file_lines, False)
                if line.strip() != '~~':
                    msg('Error', 'E005', 'Missing quest end ~~', count, line)
                cur_question = \
                    Question(subelem, group, qnum, \
                                qid, text.strip(), ans, figure, answers, fcc, \
                                pool_state.cur_group.topics)
                # Add question to Group
                pool_state.cur_group.questions.append(cur_question)
                    #case 'end':
            elif key == 'end':
                # Can end group, subelement
                pool_state.close_group()
                pool_state.close_subelement()
                pool_state.close_element()
                pool_state.state = 'end'
                pool_state.print_summary()
                    #case _:
            else:
                #print(f'Error: {key} is not valid in state "{pool_state.state}" ', end='')
                msg('Error', 'E010', f'{key} not valid', count, line)
                msg('Info', 'I003', f'{begin_state}:{pool_state.state}', count, line)
                pool_state.state = 'end'
            #case 'end':
        elif pool_state.state == 'end':
            pass
            #case _:
        else:
            pass
        if begin_state != pool_state.state:
            msg('Debug', 'D005', f'{begin_state}:{pool_state.state}', count, line)

    # If windows doc file, write out txt file
    if pool_state.cur_element.filetype == 'Microsoft Word':
        # write out text file
        #out_lines = get_file(pool_state.cur_element.filename)
        out_lines = pool_state.source_lines
        #out_lines = Filelines(out_lines)
        outpath3 = f'./output/element{pool_state.cur_element.elem}.txt'
        os.makedirs(os.path.dirname(outpath3), exist_ok=True)
        with open(outpath3, 'w', encoding='utf-8-sig') as file3:
            line_num = 0
            for line in out_lines:
                line_num += 1
                line = fix_line(line)
                if not line.isascii():
                    msg('Warning', 'W002',
                        'Non ASCII in out txt', line_num, json.dumps(line))
                file3.write(line + '\n')
        msg('Info', 'I201',
            f'text written to element{pool_state.cur_element.elem }.txt, lines={len(out_lines)}')

    return pool_state.cur_element

def main():
    """
    Execute gethamquestions if called from commandline

    """
    if len(sys.argv) >= 2:
        msg('Debug', 'D001', sys.argv[1], 0, 'nond')
        if os.path.isfile(sys.argv[1]):
            get_element_pool(sys.argv[1])
        else:
            msg('Error', 'E002', 'File not found: "' + sys.argv[1] + '"')
    else:
        msg('Error', 'E999', 'Not enough arguments')

if __name__ == '__main__':
    main()
